[#3] 회원가입 기능 구현

cyj199637 opened the pull request:
[라벨 의미]
In Progress: 해당 이슈 구현 진행 중
Feedback Wanted: 피드백 희망

구현 진행 상황과 관계 없이 Feedback Wanted 라벨이 붙어있으면 피드백 해주시면 됩니다!

cyj199637 committed:
[#3] 회원가입 기능 구현

- README 수정
- 회원가입 기능 구현
- 아이디 중복 체크 기능 구현

cyj199637 committed:
[#3] 피드백 반영

- application.properties 분리
- 중복 라이브러리 제거
- 주석으로 어노테이션에 대한 설명 추가

cyj199637 committed:
[#3] 회원가입 기능 구현 & 피드백 반영

- 비밀번호 암호화 기능 추가
- 아이디 중복 발생 시, 예외를 발생시켜 처리

f-lab-dev commented on .gitignore:
```diff
@@ -29,3 +29,6 @@ build/
 .vscode/
+
+### Database config ###
+application-db.properties
```
역할에 맡게 config 파일을 나누신 것은 좋습니다~
하지만 이 properties 파일은 profile과 관련된 작명규칙이 있습니다. 그래서 그냥 한 파일로 관리해주시는게 좋을 것 같아요~
혹은 properties 파일 이름을 다르게 해주시면 될 것 같네요. (ex. `db-secret.properties`)

f-lab-dev commented on README.md:
```diff
@@ -1 +1,18 @@
+### itda
+
+* twitter, instagram 같은 sns 서비스 구현 프로젝트
+* 기능마다 성능을 고려하여 구현하는 것을 목표로 개발
```
유지보수성도 넣어주시면 좋을 것 같아요~ 우리는 실제로 재사용성 높고 가독성 높은 코드를 지향합니다

f-lab-dev commented on README.md:
```diff
@@ -1 +1,18 @@
+
+### Application UI
+
+* https://github.com/f-lab-edu/sns-server/wiki/Application-UI
```
이런 것은 링크를 넣어주시면 좋습니다. 구글에 `마크다운 문법`을 쳐보시면 좋을 것 같아요~

`[https://github.com/f-lab-edu/sns-server/wiki/Application-UI](https://github.com/f-lab-edu/sns-server/wiki/Application-UI)`

f-lab-dev commented on pom.xml:
```diff
@@ -29,6 +29,52 @@
+        -->
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-jdbc</artifactId>
```
이 라이브러리는 기존의 라이브러리에 포함되지 않았나요?

f-lab-dev commented on pom.xml:
```diff
@@ -29,6 +29,52 @@
+        : MySQL용 JDBC Type 4 드라이버
+
+        spring-jdbc
+        : 기존 JDBC의 단점을 극복할 수 있게 해주고, 간결한 형태의 API 사용법을 제공하며,
```
용도까지 알아보신 점 좋습니다~ 그 내용도 간단하게 알아두시면 좋겠습니다. (여기 주석에 쓰시진 않아도 됩니다)

f-lab-dev replied:
spring jdbc 의존성을 명시한 곳 아래에 주석이 있어서 헷갈릴 우려가 있으니 위의 spring-jdbc 의존성 명시를 이 아래에 해주면 어떨까요?

f-lab-dev commented on pom.xml:
```diff
@@ -60,6 +117,11 @@
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
```
`spring-boot-starter-test` 안에 junit이 들어있지 않을까요?

cyj199637 replied:
네!확인해보니 `spring-boot-starter-test`에 junit5가 포함되어있었습니다. 이전에 테스트 코드를 작성할 때 @RunWith를 사용했었는데 해당 어노테이션은 junit4에서 사용되는 것이라 자동으로 라이브러리가 추가된 것 같습니다!

f-lab-dev replied:
넵~ 그리고 깃허브에서 골뱅이를 치면 진짜 사람이 태그됩니다~ 그래서 \`\`를 이용해 코드로 감싸주는게 좋습니다. `@RunWith`

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/MyBatisConfig.java:
```diff
@@ -0,0 +1,49 @@
+    * 매퍼 인터페이스 파일이 있는 가장 상위 패키지를 지정하면 지정된 패키지에서 하위 패키지를 모두 검색
+
+ @Autowired
+ : 필요한 의존 객체의 타입에 해당하는 빈을 찾아 주입
```
`@Autowired`가 아닌 빈을 주입하는 다른 어노테이션은 무엇무엇이 있을까요? 그 어노테이션과의 차이점은 무엇이 있을까요?

cyj199637 replied:
스프링 프레임워크는 @Autowired, @Inject, @Resource 어노테이션을 통해 빈을 주입해준다. 이 어노테이션들의 역할은 비슷하지만 다음과 같은 차이점이 있다.

* @Autowired

  * 스프링에서 지원하는 어노테이션
  * 빈 검색 방식
    1. byType: 먼저 명시한 타입과 일치하는 빈을 검색하고,
    2. byName: 해당 타입의 빈이 여러 개인 경우, 아이디가 일치하는 빈을 검색

  

* @Inject

  * 자바 표준 어노테이션
  * 빈 검색 방식
    1. byType: 먼저 명시한 타입과 일치하는 빈을 검색하고,
    2. byName: 해당 타입의 빈이 여러 개인 경우, 아이디가 일치하는 빈을 검색

  

* @Resource

  * 자바 표준 어노테이션
  * 빈 검색 방식
    1. byName: 먼저 명시한 빈의 아이디와 일치하는 빈을 검색하고,
    2. byType: 아이디와 일치하는 빈을 찾지 못한 경우, 타입과 일치하는 빈을 검색

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+
+        userService.signUpUser(user);
+
+        return new ResponseEntity<String>("회원가입 완료", HttpStatus.CREATED);
```
이 메소드는 항상 같은 값을 리턴하는데 계속 새로운 객체를 만들어서 리턴해주고 있습니다. 
만약 이 부분을 상수로 만들어준다면 객체 생성을 1번만 하고, 돌려쓸 수 있으니 메모리 할당을 줄일 수 있지 않을까요?
아래 부분도 마찬가지입니다~

f-lab-dev replied:
추가적으로 `CREATED`라는 응답코드에서 회원가입이 완료되었다는 점을 유추할 수 있습니다. 그렇다면 `"회원가입 완료"`라는 문자열을 굳이 body에 보내줄 필요가 있을까요?

특수한 경우에 클라이언트에 표시될 문자열을 서버에서 보내줘야할 필요가 있을 수 있습니다. 그런게 아니라면 body를 없애줘서 네트워크 대역폭을 조금이라도 덜 사용할 수 있습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+        try {
+            userService.checkIdDupe(userId);
+        } catch (NotUniqueIdException e) {
+            return new ResponseEntity<String>(e.getMessage(), HttpStatus.CONFLICT);
```
이 API를 호출하는 사람은 `중복된 아이디냐 아니냐`에 관심이 있습니다. 굳이 예외 메세지를 보여줄 필요가 있을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+        return new ResponseEntity<String>("회원가입 완료", HttpStatus.CREATED);
+    }
+
+    @GetMapping("/checkIdDupe")
```
여기는 이런식으로 가면 어떨까요? `GET /users/{username}/exists`
`id`라는 표현은 primary key 개념과 헷갈릴 수 있습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+
+        try {
+            userService.checkIdDupe(userId);
+        } catch (NotUniqueIdException e) {
```
우리는 **이 아이디가 중복된 아이디냐**라는 요청에 대한 로직을 개발하고 있습니다.
그렇다면 **아이디가 존재하지 않는다**라는 상황은 정상 동작상황일까요? 혹은 예외상황일까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/model/User.java:
```diff
@@ -0,0 +1,37 @@
+ */
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
```
이 모든 어노테이션들을 다 설정하신 이유가 무엇인가요?
모든 요소를 다 설정하지말고 필요한 것만 설정해줘야합니다. 대표적으로 immutable이 필요한 상황인데요~
`java immutable`에 대해 검색해보시면 이렇게 필요한 것만 설정해줘야하는 이유중에 하나를 알 수 있으실겁니다

cyj199637 replied:
`thread safe`, `stateless`

cyj199637 replied:
* Thread safe?

: "쓰레드에 안전하다"는 것은 멀티 쓰레드 환경에서 어떤 변수나 메소드, 객체가 여러 스레드로부터 동시에 접근이 이루어져도 프로그램의 실행에 문제가 없음을 뜻한다. 예를 들어, 하나의 쓰레드가 어떤 메소드를 호출하여 실행하고 있는 동시에 다른 쓰레드가 같은 메소드를 호출하여 실행해도 각 쓰레드에서 결과가 올바르게 나와야 한다는 것을 뜻한다.

이처럼 `Thread safe` 하게 코드를 짜려면 여러 가지 방법이 있는데 대표적으로 아래 3가지의 방식이 있다.
1. `synchronized`

: 여러 쓰레드가 같은 객체를 참조할 때, 해당 객체에 `synchronized`로 처리된 부분이 있다면 여러 쓰레드가 접근하여 그 부분을 수행하려 해도 한 번에 하나의 쓰레드만 수행할 수 있고, 다른 쓰레드들은 쓰레드가 실행을 마칠 때까지 기다려야 한다.

* `synchronized` 사용 방법
  * 메소드를 `synchronized`로 선언
  * `synchronized` 블록

2. Immutable Object

: 불변 객체란 생성 후 그 상태를 바꿀 수 없는 객체를 뜻한다. 즉, 객체가 참조하고 있는 값이 외부에서 임의로 변경되지 않는 객체이다. 특히, 멀티 쓰레드 환경에서 데이터가 불변 객체에 저장돼 있다면 여러 쓰레드에 의해서 특정 쓰레드의 데이터가 변경될 우려가 없다. 그러나 다른 값으로 변경하고 싶은 경우, 기존 값을 수정할 수 없어 계속해서 새로운 객체를 생성하기 때문에 이로 인한 성능 저하가 발생할 수 있다.

* 불변 객체 생성 방법
  * 모든 필드를 `final` 로 선언
  * 모든 필드를 `private` 로 선언
  * `setter` 를 구현하지 않는다.

3. Stateless Code

: `Stateless` 하다는 것은 이전의 상태를 기억하지 않는다는 것이다. 여기서 상태란 객체의 상태, 즉 객체의 인스턴수 변수 값을 의미한다. 예를 들어, `stateless` 한 메소드나 함수는 입력 파라미터에 영향을 받지 않고, 같은 입력 값이라면 항상 같은 결과를 리턴하는 메소드나 함수를 가리킨다. 모든 쓰레드는 이러한 메소드를 안전하게 호출할 수 있으며 다른 메소드에서 개입하거나 다른 메소드의 반환된 결과로 바뀌지 않고 예상되는 결과를 얻을 수 있다. 멀티 쓰레드 환경의 에러는 대부분 여러 쓰레드들 간에 상태를 공유하기 때문에 발생하므로, `thread-safety` 를 이루려면 `stateless` 하게 코드를 짜야한다.

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+    private UserMapper userMapper;
+
+    @Autowired
+    private PasswordEncryptor passwordEncryptor;
```
스프링은 생성자 주입을 권장하고 있습니다~
왜 그런지 주석에 달아주시면 좋을 것 같네요

cyj199637 replied:
의존성 주입을 하는 방법에는 세 가지 방법이 있다.

1. 생성자 주입
2. 수정자 주입
3. 필드 주입

스프링에서는 이 중에서도 생성자 주입 방법을 권장한다. 그 이유는 다음과 같다.

1. immutable object로 구현 가능

   : final로 선언된 변수는 생성과 동시에 초기화가 되어야 하므로 수정자 주입과 필드 주입에서는 의존 객체를 주입받을 필드를 final로 선언할 수 없다. 그러나 생성자 주입에서는 필드를 final로 선언할 수 있어 불변 객체로 생성할 수 있다.

2. `NullPointerException` 을 방지

   : 생성자 주입에서는 의존 객체를 주입하지 않으면 객체 생성 자체가 불가능하다. 즉, null을 직접 주입하지 않는 한 `NullPointerException` 이 발생하지 않는다.

3. 리팩토링을 하는데 용이

   : 주입받는 객체의 수가 늘어날수록 생성자의 인수도 늘어나 복잡한 코드가 된다. 이러한 코드는 단일 책임 원칙을 어기고 있다는 뜻이기 때문에 리팩토링의 필요성을 금방 인지할 수 있게 된다.

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public void checkIdDupe(String userId) throws NotUniqueIdException {
+        boolean isDuplicated = userMapper.checkIdDupe(userId);
```
보통 boolean을 리턴할 때에는 `is`를 앞에 붙여주는게 컨벤션입니다.
`isExistsUsername` 이런식으로 바꿔줄 수 있겠네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/tool/PasswordEncryptor.java:
```diff
@@ -0,0 +1,16 @@
+        return BCrypt.hashpw(password, BCrypt.gensalt());
+    }
+
+    public boolean match(String password, String hashedPassword) {
```
여기도 `isMatch`가 적절하겠네요~

f-lab-dev commented on src/main/resources/application.properties:
```diff
@@ -1 +1,3 @@
-
+# spring.profiles.include
+#: ����� ��ǥ�� ���е� ������� ���� ����� Ȱ��ȭ
+spring.profiles.include=db
```
profile을 사용하는 목적이 무엇일까요? 이것도 알아보시면 좋겠습니다~

cyj199637 replied:
* profile을 사용하는 이유

: 각각의 환경에 따라 설정 값이 달라지는 경우나 특정 환경에서만 쓰이는 설정이 있는 경우 사용

* 모든 환경에서 A라는 설정이 필요한데 테스트 환경에서는 A-1라는 값을 쓰고, 배포 환경에서는 A-2라는 값을 쓰고 싶다.
* B라는 설정은 ~~ 용도이므로 테스트할 때는 필요가 없고 배포할 때만 등록이 되면 좋겠다.

위와 같은 경우, 하나의 프로퍼티 파일에서 관리하면 환경이 달라질 때 마다 모든 설정 값을 바꿔야 하거나 불필요한 설정 값이 들어가게 된다. 따라서, profile을 활용하여 개발 환경에 따라 설정을 달리하고 이를 프로퍼티 파일에서 원하는 profile만 활성화하면 어플리케이션 환경 설정을 손쉽게 관리할 수 있다.

f-lab-dev commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,14 @@
+    </insert>
+
+    <select id="checkIdDupe" parameterType="String" resultType="boolean">
+        SELECT COUNT(*) > 0 FROM user
```
`COUNT(*) > 0`도 괜찮지만 다른 함수도 있을 것 같은데 찾아보시면 좋을 것 같습니다~

f-lab-dev commented on src/test/java/me/liiot/snsserver/MySQLConnectionTests.java:
```diff
@@ -0,0 +1,28 @@
+@Test
+: 해당 메소드를 테스트 대상으로 지정
+*/
+public class MySQLConnectionTests {
```
테스트를 작성하시려는 시도 좋습니다~
다만 비즈니스 로직에 대해 테스트를 작성하시려면 테스트에 대해 먼저 공부하고 작성하시는걸 추천드립니다

f-lab-dev commented:
그리고 라벨 잘 달아두셨습니다~

cyj199637 committed:
[#3] 피드백 반영

- properties 파일 분리
- pom.xml 수정
- ResponseEntity를 상수로 구현
- 아이디 중복 체크 메소드 명 및 url 수정
- User 수정
- UserMapper를 생성자로 주입하도록 수정
- 아이디 중복 체크 sql 수정

f-lab-dev commented on README.md:
```diff
@@ -3,7 +3,7 @@
 
 * twitter, instagram 같은 sns 서비스 구현 프로젝트
-* 기능마다 성능을 고려하여 구현하는 것을 목표로 개발
+* 기능마다 성능을 유지보수성을 고려하여 구현하는 것을 목표로 개발
```
`성능을 유지보수성을` 오타로 보이네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -20,25 +20,29 @@
 @RequestMapping("/users")
 public class UserController {
 
+    private final ResponseEntity RESPONSEOK = new ResponseEntity(HttpStatus.OK);
```
이런 경우는 `스네이크 케이스`를 이용하여 `RESPONSE_OK`와 같이 써주면 더 좋습니다~

cyj199637 committed:
[#3] 피드백 반영

- 상수 명 변경
- User Password 필드 변경
- UserServieceImpl 비밀번호를 암호화한 후 새로운 User로 재할당

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/UserServiceImpl.java:
```diff
@@ -24,7 +25,18 @@ public UserServiceImpl(UserMapper userMapper) {
+                user.getName(),
+                user.getPhoneNumber(),
+                user.getEmail(),
+                user.getBirth()
```
가독성을 고려해서 개행을 하셨네요~ 이렇게 길어질 경우에는 빌더 패턴을 고려해보는 것도 좋습니다

cyj199637 committed:
[#3] 피드백 반영

- User @Builder 어노테이션을 이용한 빌더 패턴 적용
- UserServiceImpl 빌더 패턴으로 User 객체 생성
- UserController ResponseEntity 제네릭 타입 변경

f-lab-dev commented:
이거는 approve 되었으므로 머지하셔도 좋습니다~
